### ISSUE TITLE
[API] Custom attributes created via the API are not valid candidates for future API created variations

### ISSUE THREAD

**Issue by kevin-frugier** [woothemes/woocommerce#11200]
## EXPLANATION OF THE ISSUE

Custom attributes created via the API (as opposed to identical attributes created via the UI) look valid but when they are later used to create variations via the API, the variations look good from a customer (purchaser) point of view but in the admin UI they look disconnected from the attribute.

![selection_242](https://cloud.githubusercontent.com/assets/9249864/16223526/053f7196-379e-11e6-9780-6a9d9e95c1a9.png)
![selection_243](https://cloud.githubusercontent.com/assets/9249864/16223527/0806d662-379e-11e6-85c0-06288ebf14cd.png)

This could be what other users complained of:
https://github.com/woothemes/woocommerce/issues/6082#issuecomment-209546361
http://stackoverflow.com/questions/37823867/woocommerce-api-create-new-product-with-variations

I first started by investigating it from a black box perspective (using the UI & the API) to narrow down the problem to the creation of the attribute and not the variations.
Then I looked at the difference between UI & API with regards to MySQL data when creating an attribute.

### Analysis from a UI & API point of view
This has many steps but shows how I managed to narrow down the problem to the creation of attributes via the API and not the creation of variations via the API which is OK.

**TL;DR:** variation creation via the API works fine, attribute creation via the API looks normal but something hidden prevents future variation creations to be properly configured

**Step 1:** Create a new product via the UI
(e.g in local : http://localhost/wp-admin/post-new.php?post_type=product)
Name : Some Product Name
Description : Some Product Description
Product Data - Variable product

**Step 2:** Fetch product via API
```
function getObject(id) {
    WooCommerce.get('products/' + id, function (err, data, res) {
        debug('res', res);
    });
}
```
Result : 
I won’t dump the full object here but basically what’s of interest to us is:
```
    "attributes": [],
    "default_attributes": [
        {
            "id": 0,
            "name": "0",
            "option": ""
        }
    ],
    "variations": [],
```

**Step 3:** Create an attribute via the UI
Custom product attribute
Name : someattribute
Value(s) : Option1
Visible on the product page : true
Used for varations : true

**Step 4:** Fetch product via API
Cf code in step 2
Apart from some date changes, the real change is the following : 
`    "attributes": [],`
became
```
    "attributes": [
        {
            "id": 0,
            "name": "someattribute",
            "position": 0,
            "visible": true,
            "variation": true,
            "options": [
                "Option1"
            ]
        }
    ],
```

**Step 5:** Create a variation via the UI
Add variation
#ID > Option 1
SKU: VARIATION-SKU
Virtual: true
Regular price: 1
Variation description: Some variation description

**Step 6:** Notice that the variation can be purchased and that it shows properly in the admin UI (after a refresh)
![selection_240](https://cloud.githubusercontent.com/assets/9249864/16223764/3a694602-379f-11e6-8976-1cb842dcac89.png)
![selection_241](https://cloud.githubusercontent.com/assets/9249864/16223767/3e702162-379f-11e6-9cf4-62305bc826cd.png)

**Step 7:** Fetch product via API
Cf code in step 2
We see the following changes
```
"price": "", → "price": 1,
"price_html": "", → "price_html": "<span class=\"woocommerce-Price-amount amount\">1,00&nbsp;<span class=\"woocommerce-Price-currencySymbol\">&euro;</span></span>",
"purchasable": false, → "purchasable": true,
"virtual": false, →  "virtual": true,
"in_stock": false, → "in_stock": true,
"shipping_required": true, → "shipping_required": false,
```
```
    "default_attributes": [
        {
            "id": 0,
            "name": "0",
            "option": ""
        }
    ],
    "variations": [],
```
Became
```
    "default_attributes": [],
    "variations": [
        {
            "id": 123,
            "date_created": "2016-06-21T06:30:34",
            "date_modified": "2016-06-21T06:36:42",
            "permalink": "http://localhost/?post_type=product&p=122&attribute_someattribute=Option1",
            "sku": "VARIATION-SKU",
            "price": "1",
            "regular_price": "1",
            "sale_price": "",
            "date_on_sale_from": "",
            "date_on_sale_to": "",
            "on_sale": false,
            "purchasable": true,
            "virtual": true,
            "downloadable": false,
            "downloads": [],
            "download_limit": -1,
            "download_expiry": -1,
            "tax_status": "taxable",
            "tax_class": "",
            "manage_stock": false,
            "stock_quantity": null,
            "in_stock": true,
            "backorders": "no",
            "backorders_allowed": false,
            "backordered": false,
            "weight": "",
            "dimensions": {
                "length": "",
                "width": "",
                "height": ""
            },
            "shipping_class": "",
            "shipping_class_id": 0,
            "image": [
                {
                    "id": 0,
                    "date_created": "2016-06-21T06:38:34",
                    "date_modified": "2016-06-21T06:38:34",
                    "src": "http://localhost/wp-content/plugins/woocommerce/assets/images/placeholder.png",
                    "name": "Placeholder",
                    "alt": "Placeholder",
                    "position": 0
                }
            ],
            "attributes": [
                {
                    "id": 0,
                    "name": "someattribute",
                    "option": "Option1"
                }
            ]
        }
    ],
```

**Step 8:** Remove the variation via the UI & fetch the product via API
It’s almost the opposite of step 7 except for `default_attributes` which remains empty
```
   "default_attributes": [],
    "variations": [],
```

**Step 9:** Create the variation via the API
The code below takes the exact dump of step 7 simply removing id, dates & permalink
```
function addVariation(productID) {
    WooCommerce.put('products/' + productID, {
        "variations": [
            {
                "sku": "VARIATION-SKU",
                "price": "1",
                "regular_price": "1",
                "sale_price": "",
                "date_on_sale_from": "",
                "date_on_sale_to": "",
                "on_sale": false,
                "purchasable": true,
                "virtual": true,
                "downloadable": false,
                "downloads": [],
                "download_limit": -1,
                "download_expiry": -1,
                "tax_status": "taxable",
                "tax_class": "",
                "manage_stock": false,
                "stock_quantity": null,
                "in_stock": true,
                "backorders": "no",
                "backorders_allowed": false,
                "backordered": false,
                "weight": "",
                "dimensions": {
                    "length": "",
                    "width": "",
                    "height": ""
                },
                "shipping_class": "",
                "shipping_class_id": 0,
                "image": [
                    {
                        "id": 0,
                        "src": "http://localhost/wp-content/plugins/woocommerce/assets/images/placeholder.png",
                        "name": "Placeholder",
                        "alt": "Placeholder",
                        "position": 0
                    }
                ],
                "attributes": [
                    {
                        "id": 0,
                        "name": "someattribute",
                        "option": "Option1"
                    }
                ]
            }
        ]
    }, function (err, data, res) {
        debug('err', format(err));
        debug('data', format(data));
        debug('res', format(res));
    });
}
```

**Step 10:** Notice that the variation can be purchased and that it shows properly in the admin UI (after a refresh)
Screens identical to step 6 with a different ID

**Step 11:** Fetch product via API
We see the same results as step 7 except for `default_attributes` which here was already `[]` and remains so.

**Step 12:** Remove both the variation and the attribute via the UI & fetch product via API
We are back to a basic product
```
{
    "id": 122,
    "name": "Some Product Name",
    "slug": "",
    "permalink": "http://localhost/?post_type=product&p=122",
    "date_created": null,
    "date_modified": "2016-06-21T07:08:04",
    "type": "variable",
    "status": "draft",
    "featured": false,
    "catalog_visibility": "visible",
    "description": "<p>Some Product Description</p>\n",
    "short_description": "",
    "sku": "",
    "price": "",
    "regular_price": "",
    "sale_price": "",
    "date_on_sale_from": "",
    "date_on_sale_to": "",
    "price_html": "",
    "on_sale": false,
    "purchasable": false,
    "total_sales": 0,
    "virtual": false,
    "downloadable": false,
    "downloads": [],
    "download_limit": -1,
    "download_expiry": -1,
    "download_type": "standard",
    "external_url": "",
    "button_text": "",
    "tax_status": "taxable",
    "tax_class": "",
    "manage_stock": false,
    "stock_quantity": null,
    "in_stock": false,
    "backorders": "no",
    "backorders_allowed": false,
    "backordered": false,
    "sold_individually": false,
    "weight": "",
    "dimensions": {
        "length": "",
        "width": "",
        "height": ""
    },
    "shipping_required": true,
    "shipping_taxable": true,
    "shipping_class": "",
    "shipping_class_id": 0,
    "reviews_allowed": true,
    "average_rating": "0.00",
    "rating_count": 0,
    "related_ids": [],
    "upsell_ids": [],
    "cross_sell_ids": [],
    "parent_id": 0,
    "purchase_note": "",
    "categories": [],
    "tags": [],
    "images": [
        {
            "id": 0,
            "date_created": "2016-06-21T07:08:22",
            "date_modified": "2016-06-21T07:08:22",
            "src": "http://localhost/wp-content/plugins/woocommerce/assets/images/placeholder.png",
            "name": "Placeholder",
            "alt": "Placeholder",
            "position": 0
        }
    ],
    "attributes": [],
    "default_attributes": [],
    "variations": [],
    "grouped_products": [],
    "menu_order": 0,
    "_links": {
        "self": [
            {
                "href": "http://localhost/wp-json/wc/v1/products/122"
            }
        ],
        "collection": [
            {
                "href": "http://localhost/wp-json/wc/v1/products"
            }
        ]
    }
}
```

**Step 13:** Create the attribute via the API
Created with the following code which simply puts back what was fetched in step 4
```
function addAttribute(productID) {
    WooCommerce.put('products/' + productID, {
        "attributes": [
            {
                "id": 0,
                "name": "someattribute",
                "position": 0,
                "visible": true,
                "variation": true,
                "options": [
                    "Option1"
                ]
            }
        ]
    }, function (err, data, res) {
        debug('err', format(err));
        debug('data', format(data));
        debug('res', format(res));
    });
}
```

**Step 14:** Fetch product via API
We see the **same result** as step 4 except for `default_attributes` which here was already `[]` and remains so.

**Step 15:** Create the variation via the API
Repeat step 9

**Step 16:** Notice that the variation can be purchased **but** that it **doesn’t show properly** in the admin UI (after a refresh)
Same picture for the product purchase as step 6
But the view in the admin shows the variation is not properly bound to the attribute value (even though from the customer view in the purchase it looks good)
![selection_242](https://cloud.githubusercontent.com/assets/9249864/16223526/053f7196-379e-11e6-9780-6a9d9e95c1a9.png)
The dropdown value can’t even be fixed manually : 
![selection_243](https://cloud.githubusercontent.com/assets/9249864/16223527/0806d662-379e-11e6-85c0-06288ebf14cd.png)

**Step 17:** Fetch product via API
We see the exact same result as step 11 & 7 except for some date & id changes (since it’s a new variation)

**So:** the API doesn’t allow us to see where the problem is in term of data :warning: 

**Step 18:** Remove both the variation and the attribute via the UI & fetch product via API
We are back to step 12

**Step 19:** Create the attribute via the UI and the variation via the API
Repeat step 3 & step 9/15

**Step 20:** Notice that the variation can be purchased and that it shows properly in the admin UI (after a refresh)
Same as step 6 everything is OK which confirms there is no problem with variations but only with custom attribute creation via the API (problem not detectable when fetching objects via the API because problematic objects look the same as good objects).

**Step 21:** Fetch product via API
It looks like steps 7 & 14

### Analysis with MySQL
Now that it has been identified that there is something different between the creation of a custom attribute via the UI & via the API which prevents variations to work, I’ll compare the two in MySQL.

**TL;DR:** `position` or `is_taxonomy` are not handled similarly when creating an attribute from the API and from the UI

I use the following query to fetch the properties of the product
`SELECT * FROM wordpress.wp_postmeta WHERE post_id=122;`

**MySQL:** Creation of the custom attribute via the UI (same as step 3)
`1566,122,_product_attributes,"a:1:{s:13:"someattribute";a:6:{s:4:"name";s:13:"someattribute";s:5:"value";s:7:"Option1";s:8:"position";s:1:"0";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:0;}}"`

**API**: Creation of the custom attribute via the API (same as step 13)
`1566,122,_product_attributes,"a:1:{s:13:"someattribute";a:6:{s:4:"name";s:13:"someattribute";s:5:"value";s:7:"Option1";s:8:"position";i:0;s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";b:0;}}"`

We see a difference in the attribute `position`: when created via the UI it is a `string` but when fetch the object via the API it is an `integer` so no surprise when we create it with an integer it is stored as such.
Then we see also a difference in `is_taxonomy`: when created via the UI it is an `integer`, when we create via the API (even though we don’t give this attribute in the query) it is a `boolean`

So, I tuned the way I create the attribute via the API:
```
function addAttribute(productID) {
    WooCommerce.put('products/' + productID, {
        "attributes": [
            {
                "id": 0,
                "name": "someattribute",
                "position": "0",
                "visible": true,
                "variation": true,
                "options": [
                    "Option1"
                ],
                "is_taxonomy": 0
            }
        ]
    }, function (err, data, res) {
        debug('err', format(err));
        debug('data', format(data));
        debug('res', format(res));
    });
}
```

But after inspecting in MySQL we see the same as before. Even though we explicitly passed a string for position and an integer for is_taxonomy, they are back to integer & boolean (as opposed to a creation via the UI).

:white_check_mark: If I change it manually in MySQL to what it should would be if I had created it via the UI (string & boolean), **then** when I create the variation via the API everything works fine.

:warning: **So** the problem lies in the way either `position` or `is_taxonomy` is handled in the API.

## SYSTEM STATUS REPORT

```
WordPress Environment
Home URL:		http://localhost
Site URL:		http://localhost
WC Version:		2.6.1
Log Directory Writable:		 /var/www/html/wp-content/uploads/wc-logs/
WP Version:		4.5.2
WP Multisite:		
WP Memory Limit:		256 MB
WP Debug Mode:		–
WP Cron:		
Language:		en_US
Server Environment
Server Info:		Apache/2.4.10 (Debian) PHP/5.6.21
PHP Version:		5.6.21
PHP Post Max Size:		8 MB
PHP Time Limit:		30
PHP Max Input Vars:		1000
cURL Version:		7.38.0, OpenSSL/1.0.1k
SUHOSIN Installed:		–
MySQL Version:		5.7.11
Max Upload Size:		2 MB
Default Timezone is UTC:		
fsockopen/cURL:		
SoapClient:		 Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument:		
GZip:		
Multibyte String:		
Remote Post:		
Remote Get:		
Database
WC Database Version:		2.6.1
woocommerce_sessions	 	
woocommerce_api_keys	 	
woocommerce_attribute_taxonomies	 	
woocommerce_downloadable_product_permissions	 	
woocommerce_order_items	 	
woocommerce_order_itemmeta	 	
woocommerce_tax_rates	 	
woocommerce_tax_rate_locations	 	
woocommerce_shipping_zones	 	
woocommerce_shipping_zone_locations	 	
woocommerce_shipping_zone_methods	 	
woocommerce_payment_tokens	 	
woocommerce_payment_tokenmeta	 	
MaxMind GeoIP Database:		 The MaxMind GeoIP Database does not exist - Geolocation will not function. You can download and install it manually from http://dev.maxmind.com/geoip/legacy/geolite/ to the path: /var/www/html/wp-content/uploads/GeoIP.dat
Active Plugins (0)
WooCommerce	 	by WooThemes – 2.6.1
HookPress	 	by mitcho (Michael Yoshitaka Erlewine) – 1.14
Settings
Force SSL:		–
Currency		EUR (€)
Currency Position		right_space
Thousand Separator		
Decimal Separator		,
Number of Decimals		2
API
API Enabled:		
WC Pages
Shop Base:		#40 - /shop/
Cart:		#41 - /cart/
Checkout:		#42 - /checkout/
My Account:		#43 - /my-account/
Taxonomies
Product Types:	 	external (external), grouped (grouped), simple (simple), variable (variable)
Theme
Name:		Divi
Version:		2.7.5
Author URL:		http://www.elegantthemes.com
Child Theme:		 – If you're modifying WooCommerce on a parent theme you didn't build personally, then we recommend using a child theme. See: How to create a child theme
WooCommerce Support:		
Templates
Overrides:	 	–
```

**Comment by claudiosmweb** [comment]
Hard to understand it, since is too big and a lot of steps that don't makes any sense.
I saw one bug about `default_attributes` and I'm fixing, but not sure what more is happening here.

**Comment by claudiosmweb** [comment]
Wow, finally I understand the other bug :tada: :fireworks:

**Comment by kevin-frugier** [comment]
@claudiosmweb : indeed that's a very long issue description but, for my defense, I added several **TL;DR** along the way especially for fast readers.

That's great seeing the trouble I went through to narrow down the issue is considered to make "no sense"...
To clarify : they are not steps meant to be executed one by one by anyone trying to reproduce the issue. They are steps showing the logical path I went through (removing/adding stuff via UI, then via API). I put myself in the shoes of the WooCommerce user and try to be 100% sure the bug I'm seeing is due to a bug (hence created many steps around the issue with several approaches) and not a bad usage of the API.

Now, let me extract the bits of information that are relevant because I'm afraid you missed the point of the issue and might not fix it even though it's pretty clear what is wrong:

**Important stuff start**
From TL;DR #1: variation creation via the API works fine, attribute creation via the API looks normal but something hidden prevents future variation creations to be properly configured
From TL;DR #2: `position` or `is_taxonomy `are not handled similarly when creating an attribute from the API and from the UI

To summarize : 
From the **UI**, attribute `position `is handled as a string and `is_taxonomy ` as an integer. This makes future variations created from the API **work** properly.
From the **API**, attribute `position `is handled as an integer and `is_taxonomy ` as an boolean. This makes future variations created from the API **break** miserably.

:warning:  and that is not related to `default_attributes`

**Important stuff end**

**Comment by claudiosmweb** [comment]
@kevin-frugier Thanks for your help, but next time try to be more direct...

**Comment by claudiosmweb** [comment]
@kevin-frugier I take more than one hour to fix the type of 3 variables... This is not cool for me or your.

**Comment by claudiosmweb** [comment]
> That's great seeing the trouble I went through to narrow down the issue is considered to make "no sense"...
>To clarify : they are not steps meant to be executed one by one by anyone trying to reproduce the issue. They are steps showing the logical path I went through (removing/adding stuff via UI, then via API). I put myself in the shoes of the WooCommerce user and try to be 100% sure the bug I'm seeing is due to a bug (hence created many steps around the issue with several approaches) and not a bad usage of the API.

Ok, next time you can try "Hi guys, I found 3 bugs on Products REST API, the first one is with _default_attributes return an array of empty values, and `is_taxonomy` should save as `bool` and `position` as `string`, both `is_taxonomy` and `position` are causing bugs on the admin screen".

This is more easy to understand or fix, or since you debugged, you can submit a Pull Request too.
Too many steps make people confuse and make hard to see what is going wrong.

**Comment by kevin-frugier** [comment]
@claudiosmweb : indeed I could save you time in the future putting the summary on top.
The rest of my post was more here as a way to right down all the side information that lead to the understanding of the bug. That's an old work habit from a previous job (not related to opensource & GitHub obviously :laughing: ) where bugs had to be thoroughly explained before even being considered fixable.
I'll try to be nicer to you on future issues (but I hope there won't be many :smile: )

**Comment by claudiosmweb** [comment]
@kevin-frugier sorry man, I understand that, but no need to be too fancy while opening an issue here.
You can said only "this s*** is broken" and I'll test and fix, no worry :)

**Comment by claudiosmweb** [comment]
@kevin-frugier and please, if you feel comfortable, please send Pull Requests, so we have you on our list of contributors

**Comment by kevin-frugier** [comment]
@claudiosmweb I'll consider that on the next issue (depending on how easy it is to fix).

Thanks for your reactivity anyway, I can't wait to start playing again with my webhooks & the API now that you fixed it.
